bundles/org.openhab.binding.gree/src/main/java/org/openhab/binding/gree/internal/discovery/GreeDiscoveryService.java

```diff
@@ -0,0 +1,130 @@
+import org.eclipse.smarthome.core.net.NetworkAddressService;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.gree.internal.GreeException;
+import org.openhab.binding.gree.internal.GreeTranslationProvider;
+import org.openhab.binding.gree.internal.handler.GreeAirDevice;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link GreeDiscoveryService} implements the device discovery service. UDP broadtcast ius used to find the devices on
+ * the local subnet.
+ *
+ * @author Markus Michels - Initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(service = GreeDiscoveryService.class, immediate = true, configurationPid = "discovery.gree")
+public class GreeDiscoveryService extends AbstractDiscoveryService {
+    private static final int TIMEOUT_SEC = 10;
+    private final Logger logger = LoggerFactory.getLogger(GreeDiscoveryService.class);
+    private final GreeTranslationProvider messages;
+    private final String broadcastAddress;
+
+    private GreeDeviceFinder deviceFinder = new GreeDeviceFinder();
+
+    @Activate
+    public GreeDiscoveryService(@Reference NetworkAddressService networkAddressService,
+            @Reference LocaleProvider localeProvider, @Reference TranslationProvider i18nProvider) {
+        super(SUPPORTED_THING_TYPES_UIDS, TIMEOUT_SEC);
+        Bundle bundle = FrameworkUtil.getBundle(this.getClass());
+        messages = new GreeTranslationProvider(bundle.getBundleContext().getBundle(), i18nProvider, localeProvider);
+        String ip = networkAddressService.getConfiguredBroadcastAddress();
+        broadcastAddress = ip != null ? ip : "";
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        startScan();
+    }
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        stopScan();
+    }
+
+    @Override
+    protected void startScan() {
+        Optional<DatagramSocket> clientSocket = Optional.empty();
+        try {
+            clientSocket = Optional.of(new DatagramSocket());
+            deviceFinder = new GreeDeviceFinder(broadcastAddress);
+            deviceFinder.scan(clientSocket, true);
+
+            int count = deviceFinder.getScannedDeviceCount();
+            logger.info("{}", messages.get("discovery.result", count));
+            if (count > 0) {
+                logger.debug("Adding uinits to Inbox");
+                createResult(deviceFinder.getDevices());
+            }
+        } catch (GreeException e) {
+            logger.warn("Discovery failed: {}", e.toString());
+        } catch (IOException | RuntimeException e) {
+            logger.debug("Discovery failed", e);
```

Comment:
Do you log a stack trace by intention, when an IOException raises? RuntimeExceptions should be logged to warn or error. Logging IOException to debug is fine.

Reply:
changed to 
```
       } catch (GreeException e) {
            logger.warn("Discovery failed: {}", e.toString());
        } catch (IOException e) {
            logger.debug("Discovery failed", e.toString());
        } catch (RuntimeException e) {
            logger.warn("Discovery failed", e);
        } finally {
 ```